go-app/server/security.go

```diff
@@ -0,0 +1,104 @@
+package main
+
+import (
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"hash"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+)
+
+// Signature headers
+const (
+	hSignature = "X-Slack-Signature"
+	hTimestamp = "X-Slack-Request-Timestamp"
+)
+
+// SecretsVerifier contains the information needed to verify that the request comes from Slack
+type SecretsVerifier struct {
+	signature []byte
+	hmac      hash.Hash
+}
+
+func unsafeSignatureVerifier(header http.Header, secret, body string) (SecretsVerifier, error) {
+	var (
+		bsignature []byte
+	)
+
+	signature := header.Get(hSignature)
+	stimestamp := header.Get(hTimestamp)
+
+	if signature == "" || stimestamp == "" {
+		return SecretsVerifier{}, errors.New("missing headers")
+	}
+
+	bsignature, err := hex.DecodeString(strings.TrimPrefix(signature, "v0="))
+
+	if err != nil {
+		return SecretsVerifier{}, err
+	}
+
+	hash := hmac.New(sha256.New, []byte(secret))
+	basestring := fmt.Sprintf("v0:%s:%s", stimestamp, body)
```

Comment:
In the original file they weren't including the request body in this base string, it was just `fmt.Sprintf("v0:%s:", stimestamp)`. So at least in my case it would always fail to verify - which makes sense now that I've read this source code.

Reply:
Alright, turns out I wasn't thinking about this in a galaxy-brain enough way. SecretsVerifier implements Writer, which means you can write your request body string to the hash after you've created it 🤦 So I'm now going to do it like that. I think I may add a doc comment to this struct or something that tells you the intended usage.